### PR TITLE
fix(813): @Nat→@Int widening soundness (+ #819 doc-gate, v0.0.186)

### LINKED ISSUE #813: verifier: @Nat → @Int widening of a value > i64.MAX is unsound (Tier-1 proves a false postcondition)
Found while fixing #812 (PR #811). A **distinct, more serious** soundness hole than #812: this is not about literals, and it is **not codegen-guarded** — it silently runs and violates a proved contract.

`@Nat` is u64 and `@Int` is i64, and `@Nat <: @Int`.  But a `@Nat` value in (i64.MAX, u64.MAX] reinterprets its bit pattern when widened to `@Int` (e.g. u64.MAX → -1), while the verifier reasons over the unbounded non-negative mathematical value.  So Tier-1 proves postconditions the runtime then violates.

Repro:
```vera
public fn widen(@Nat -> @Int)
  requires(true) ensures(@Int.result >= 0) effects(pure)
{ @Nat.0 }
```
```
$ vera verify widen.vera          # ensures(@Int.result >= 0) verifies CLEAN at Tier 1
$ vera run widen.vera --fn widen -- 18446744073709551615
Error: Postcondition violation in widen(@Nat -> @Int)
  ensures(@Int.result >= 0) failed
```

The verifier proves `result >= 0` (a `@Nat` is non-negative), but the runtime returns `-1` for the u64.MAX input — a Tier-1/runtime divergence, exactly the #392 class.

**Scope vs #812:** #812 (fixed in #811) range-checks integer **literals** at check time.  This issue is about **non-literal `@Nat` values** crossing the `@Nat → @Int` boundary — the subtype relation, not a literal.  The negated-literal corner `-18446744073709551615` (magnitude in the u64 band) is one manifestation that #812's literal check also does not cover.

**Fix direction (a focused #392-style sub-task, like #798/#799):** emit a verifier obligation at every `@Nat → @Int` coercion site that the value is `<= i64.MAX` — provably safe → Tier 1; provably unsafe → loud error; dynamic → runtime-guarded Tier 3 — with a codegen guard mirror and a verify-vs-run differential.  (Alternatively reconsider whether `@Nat <: @Int` should be implicit at all.)

Parent: #392 (the `smt.py`/verifier↔runtime soundness audit).

### LINKED ISSUE #819: Gate: type-check doc user-example blocks against E151 (exclude spec/09 reference signatures)
## Problem

The documentation example validators (`scripts/check_skill_examples.py`, `scripts/check_spec_examples.py`, and the other `check_*_examples.py` scripts) only **parse** Vera code blocks — they do not run `vera check`. So a doc code block that defines a function named after a built-in (e.g. `fn clamp`, `fn abs`, `fn sign`) parses cleanly but would **fail `vera check` with E151** (redefining a built-in, added in #817 / v0.0.185). Such broken examples drift into the docs undetected by CI.

This bit us during #817: after E151 landed, `SKILL.md` and `DE_BRUIJN.md` still contained illustrative examples defining `clamp` / `abs` / `sign`, all of which now E151. They were found by a manual audit, not by a gate — which is exactly the kind of thing a gate should catch.

## Why not just `vera check` every doc block

`spec/09-standard-library.md` legitimately documents built-in **signatures** as reference material: `fn abs(@Int -> @Int) ...`, `fn min(...)`, `fn string_*(...)`, etc. Those blocks are *supposed* to look like definitions of built-in names; running `vera check` on them would (correctly) E151. So the gate must distinguish:

- **user-example** blocks — illustrative programs a reader would write; these MUST pass `vera check` (no E151, no other diagnostics).
- **reference-signature** blocks — the stdlib reference in `spec/09`; these are exempt (parse-only).

## Proposed approach

- Add a check-level validation pass for doc user-example blocks: run `vera check` (not just parse) and fail on any diagnostic, including E151.
- Exempt the `spec/09` reference-signature blocks via an explicit, reviewable mechanism — e.g. an allowlist keyed by (file, block index) or, preferably, a fenced-block annotation / HTML-comment marker the validator recognizes as "reference signature, parse-only."
- Wire the gate into the relevant `check_*.py` script(s) and into pre-commit + CI.

## Acceptance criteria

- A doc user-example block that defines a built-in name (or otherwise fails `vera check`) makes the gate **fail**.
- `spec/09` reference-signature blocks remain exempt and the gate is green on the current tree.
- Pre-commit and CI run the gate.

## Context

Root-caused during #817 (v0.0.185, the E151 built-in-redefinition work). Per maintainer, this lands alongside #813 in the next PR rather than as a standalone release.

### PR DESCRIPTION
**Pull request by aallan**
## Summary

Closes the `@Nat` → `@Int` widening soundness hole found during the #807 review,
and ships the #819 documentation built-in-redefinition gate alongside it.
Cuts **v0.0.186**.

`@Nat` is an unsigned i64 and `@Int` a signed i64 with `@Nat <: @Int`, so a
`@Nat` in `(i64.MAX, u64.MAX]` *bit-reinterprets* to a negative `@Int` when
widened (`u64.MAX` → `-1`). The verifier reasoned over the mathematical
non-negative value, so it could **prove a false Tier-1 postcondition** the
runtime then violated:

```vera
public fn widen(@Nat -> @Int)
  requires(true) ensures(@Int.result >= 0) effects(pure)
{ @Nat.0 }
```

`vera verify` proved `@Int.result >= 0`, yet `vera run widen(18446744073709551615)`
returned `-1`.

## It's not one site — it's ten

A 3-surface completeness audit (codegen sites, verifier obligation sites, and
the language grammar), each finding cross-checked with an empirical
`vera run -- 18446744073709551615` probe, found the coercion happens at **ten**
distinct sites. The fix handles each according to what code generation can prove
at the site:

**Runtime-trapped** (codegen can statically determine the source is `@Nat`, so it
emits a guard that traps on `u64.MAX` instead of returning the reinterpreted
value):

- return position
- `let` binding
- call argument
- concrete `@Int` constructor field
- `@Nat`-field ADT sub-pattern extraction
- match binding

**Disclosed as an `E531` warning** (code generation architecturally cannot guard
the component coercion — no per-element / monomorphized guard site — so the
verifier flags it as unguarded rather than a silent false Tier-1):

- tuple construction
- tuple destructure
- array-literal element
- generic-instantiated `@Int` field (e.g. `Some(@Nat.0)` into `Option<Int>`)

**Verifier obligation only** (code generation already covers it via desugaring):

- pipe argument (`@Nat.0 |> f()` desugars to the guarded call argument)

## Design

The verifier emits a `nat_to_int_coerce` obligation (**E530**) at every coercion
site, discharged with the two-check pattern: provably `<= i64.MAX` → Tier 1;
provably out of range → a loud `E530` error; otherwise Tier-3. Tier-3 then splits
on whether code generation guards the site: `tier3_runtime` (a runtime trap) or
`tier3_unguarded` (the new **E531** disclosure warning), mirroring the `E503` /
`E504` split on the `@Int` → `@Nat` narrowing side.

The crux that makes this correct: **a widen guard traps on a negative i64, so it
fires only when the source is provably `@Nat`** — never on a genuine `@Int`,
which may be legitimately negative. A control test confirms an `@Int` match-bind
of `-5` returns `-5` without trapping.

A verifier↔codegen behavioural **differential**
(`tests/test_int_widening_differential.py`) pins the soundness contract: every
codegen-guarded site traps on `u64.MAX`, and every `E531`-disclosed site does
not — so a verifier deferral code generation never guards (an unsound silent -1)
or a spurious trap would turn the suite red.

## Also in this PR

- **Spec correction.** spec/02-types.md §2.2.1 and SKILL.md both claimed
  `@Nat` → `@Int` widening was "unconditionally safe — no proof obligation, no
  runtime check" — the exact false premise this fix corrects. Both now
  distinguish the *type*-level subtyping rule from the *value*-level machine
  coercion, and spec/11-compilation.md §11.2.1 documents the widening guard.
- **Conformance.** `tests/conformance/ch04_nat_int_widening.vera` — the positive
  dual of `ch04_nat_binding`, with bounded widenings that discharge at Tier 1.
- **#819 doc-gate.** `scripts/check_doc_builtin_shadowing.py` (CI) verifies no
  documentation example redefines an opaque verifier-modelled built-in, so a doc
  snippet can never reintroduce the v0.0.185 E151 `verify`↔`run` divergence.

## Deferred

The remaining coercion sites — effect-operation argument, closures, heterogeneous
`if`/`match` arms, codegen-guarding the currently-`E531`-disclosed component
sites, and the explicit `nat_to_int` built-in (which models the conversion as
identity and is itself unsound on `u64.MAX`) — are tracked in #820.

Closes #813
Closes #819

## Summary by CodeRabbit

* **New Features**
  * Added full verifier- and codegen support for `@Nat`→`@Int` widening, including the new widening obligation and required runtime checks.
  * Extended widening handling across returns, call/pipe arguments, `let`s, constructors, and match/binding/destructuring patterns.

* **Bug Fixes**
  * Improved out-of-range conversion safety with correct verification errors/warnings and trapping/disclosure behaviour.

* **Tests**
  * Added regression, differential, and stage-3 codegen tests for issue coverage, plus documentation validation for built-in redefinition (E151).

* **Chores**
  * Updated conformance/testing documentation and counts, and bumped the release version to `0.0.186`.

## Follow-up: two more widening sites fixed, deferral reconciled

After the initial audit, an adversarial re-audit of every site originally deferred to #820 (each with a `vera run -- 18446744073709551615` repro) found two were fixable *now*, not architecturally blocked:

- **The explicit `nat_to_int(@Nat)` built-in.** Its declared `@Int` return masked the `@Nat` source, so `vera verify` falsely proved `ensures(@Int.result >= 0)` while `vera run -- 18446744073709551615` returned `-1`. Both `_result_is_nat` helpers now special-case it, so it carries the `nat_to_int_coerce` obligation + the boundary trap guard. **This caught four real unguarded widenings in shipping examples** (`json.vera`, `life.vera`).
- **Heterogeneous `if`/`match` with a non-negative-literal alternative** (`if c then { @Nat.0 } else { 0 }`). A literal arm read as not-`@Nat`, so the all-arms-`@Nat` boundary check mis-classified the whole expression `@Int` and the real `@Nat` arm widened unguarded. `_result_is_nat` now treats a non-negative literal arm as `@Nat`-compatible (always `<= i64.MAX`, so it can neither out-of-range-widen nor false-trap), keeping the expression `@Nat` so the single boundary guard covers the real `@Nat` arm.

The remaining deferred sites are genuinely architecture-blocked — code generation has no per-component target-type metadata (empirically confirmed: the checker's side-tables record nothing for if/match arms or tuple/array components, and `compile()` is not even handed the target-type table) — the same blocker that defers the `@Int → @Nat` *narrowing* duals. They've been reconciled out of the standalone #820: the effect-op / generic-field / tuple widenings now ride with their narrowing duals in #754 / #757 / #758, and the array-element, closure, and genuine-`@Int`-slot heterogeneous-arm widenings stay in #820. `KNOWN_ISSUES.md` gains a `@Nat → @Int` widening Limitations row (the dual of the narrowing row).

Tests are test-first (RED on the pre-fix verify+run): `nat_to_int` and heterogeneous-literal trap / no-trap / E530 codegen cases, an `@Int`-arm no-false-trap control (mutation-validated), verifier-obligation status tests, and three verifier↔codegen differential corpus entries. `test_overall_tier_counts` moves 271/88/359 → 271/92/363 (the +4 are the `nat_to_int` widenings now obligated in `json`/`life`).

### CODE REVIEW

**Comment by aallan** [comment]
## Review pass (pr-review-toolkit)

Ran three reviewers over the branch diff — code quality, test coverage, and silent-failure hunting. All three independently converged on one **critical** residual hole, now fixed; everything else was confirmed sound.

### Critical (fixed)

A `@Nat`-returning **call result** widened to `@Int` was classified `tier3` (runtime-guarded) by the verifier but emitted **no** codegen guard — `vera verify` reported "1 runtime check (Tier 3)" while `vera run … -- 18446744073709551615` silently returned `-1`. The original soundness hole, relabelled. It affected return / `let` / call-argument sites whose value is a call result (concrete or `forall<T>` monomorphised to `@Nat`); the differential missed it because its corpus only used bare-slot values.

Two root causes, both fixed:

1. The verifier and codegen `_result_is_nat` disagreed for calls. Codegen used `_infer_fncall_vera_type`, which maps the i64 WASM return back to `"Int"` and can never yield `"Nat"` (both lower to i64). It now reads the checker's resolved-type side-table (`_resolved_codegen_type`), matching the verifier, with the inference helper as a fallback for built-in `@Nat` returns.
2. The return-position guard is appended after the trailing tail call, so `return_call` returned before it ran. Codegen now reverts `return_call`→`call` for `@Nat`→`@Int` returns (the same revert the postcondition-check path uses).

All return / `let` / call-arg call-result sites now trap on `u64.MAX` and round-trip in-range values.

### Test hardening

- `test_int_widening_differential.py` `_GUARDED` gains the call-result cases (concrete + generic-instantiated) and an `if`-join case — the desync the corpus had been blind to.
- Codegen controls: call-result-return trap, return/`let` `@Int`-source no-trap (the guard keys on a `@Nat` source, never a genuine `@Int`), pipe-arg trap.
- A non-return `E530` test, and the `tier3_unguarded` docstring now lists `E531`.

### Confirmed sound (no change needed)

- The `int_fields`/`nat_fields` layout bitmaps are disjoint and length-validated; the ADT sub-pattern guard keys on the *source* field type, so it can't fire on a genuine `@Int` field.
- A genuine negative `@Int` through return / `let` / match-bind round-trips without trapping.
- The `_check_int_widening_obligation` total bookkeeping is balanced, mirroring the overflow/nat-bind obligations.

Deferred coercion sites (effect-op argument, closures, heterogeneous arms, the component coercions disclosed as `E531`, and the explicit `nat_to_int` built-in) remain tracked in #820.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -89,7 +89,7 @@ dependencies. CI enforces that `uv.lock` stays current.
...
+Every push is checked by 29 configured hooks across two stages: 27 are configured at the commit stage (run after `pre-commit install`), and 2 (`check-changelog-updated` and `uv-lock-check`, described below) are configured at the push stage (run after `pre-commit install --hook-type pre-push`). Most commit-stage hooks have per-hook `files:` / `types:` filters — the `python` type-check only runs when Python files are staged; `check_readme_examples.py` only runs when `README.md` or Vera sources change, etc. A plain-text commit touching only one markdown file triggers a small subset; a compiler-level commit triggers most of them.
 
 The **commit-stage** hooks (26, each gated to relevant files) include:
```
[the remark this answers is not in the document]
Comment (aallan) [reply CONTRIBUTING.md]: @coderabbitai Fixed — bumped the commit-stage hook breakdown from 26 to 27 to match the summary line (the #819 doc-shadowing hook added one).

File: scripts/check_doc_builtin_shadowing.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+import sys
+from pathlib import Path
+
+from vera.environment import TypeEnv
+from vera.prelude import overridable_builtin_names
+
+# A `fn` definition (top-level, module, or `where`-block); `public`/`private`
+# optional. This def form only appears in code blocks — prose references carry
+# backticks ("the `fn abs`"), not a bare line-leading `fn abs(`.
+_FN_DEF = re.compile(r"^\s*(?:public |private )?fn\s+([a-z_][A-Za-z0-9_]*)\b")
```
[the remark this answers is not in the document]
Comment (aallan) [reply scripts/check_doc_builtin_shadowing.py]: @coderabbitai Fixed — `_FN_DEF` now matches an optional `forall<...>` generic header (including a `where`-bounded one with nested `<...>`), with a regression test in test_doc_builtin_shadowing.py.

File: scripts/check_doc_builtin_shadowing.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    return hits
+
+
+def doc_files(root: Path) -> list[Path]:
+    """The doc surfaces to scan: top-level ``*.md`` (README, SKILL, DE_BRUIJN,
+    FAQ, …) plus ``spec/*.md``. Generated (``docs/``), artefact (``mutants/``),
+    and dependency (``.venv/``) trees are excluded by construction — they are
+    not direct children here."""
+    files = sorted(root.glob("*.md")) + sorted((root / "spec").glob("*.md"))
+    return [f for f in files if f.relative_to(root).as_posix() not in _EXEMPT]
```
[the remark this answers is not in the document]
Comment (aallan) [reply scripts/check_doc_builtin_shadowing.py]: @coderabbitai Fixed — `doc_files()` now scans all authored `*.md` via `rglob` (including examples/README.md), excluding generated/vendored/artefact trees; the gate still passes (141 built-ins across 45 doc files).

File: tests/test_verifier.py [hunk 4]
```diff
@@ -3514,7 +3514,13 @@ def test_tier_counts_with_imports(self) -> None:
...
         # requires(true) → Tier 1, ensures(@Int.result >= 0) → Tier 1 (via abs postcondition)
         assert result.summary.tier1_verified >= 2
-        assert result.summary.tier3_runtime == 0
+        # #813: `abs` returns @Nat, so `wrap`'s `{ abs(@Int.0) }` widens a @Nat
+        # result into the @Int return — a nat_to_int_coerce obligation.  The
+        # value can exceed i64.MAX (abs(i64.MIN) = 2**63), so it is honest Tier-3
+        # (runtime-guarded), not provable.  This is exactly the soundness hole
+        # #813 closes: without it `ensures(@Int.result >= 0)` proved over the
+        # unbounded @Nat while abs(i64.MIN) returns a negative i64 at runtime.
+        assert result.summary.tier3_runtime == 1
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_verifier.py]: @coderabbitai Fixed — removed that assertion. The imported `math.abs` is declared `@Int -> @Int`, so the tier3 count came from the bare `abs` resolving to the @Nat built-in (an import-vs-built-in edge case), not this test purpose; the #813 widening is covered by the dedicated suites.

File: tests/test_int_widening_differential.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    # side-table (`_infer_fncall_vera_type` maps the erased i64 back to "Int"),
+    # at both the return and let sites.  Pre-fix these returned -1.
+    ("generic_return", """
+private forall<T> fn identity(@T -> @T) requires(true) ensures(true) effects(pure) { @T.0 }
+public fn f(@Nat -> @Int) requires(true) ensures(true) effects(pure) { identity(@Nat.0) }
+""", "f"),
+    ("generic_let", """
+private forall<T> fn identity(@T -> @T) requires(true) ensures(true) effects(pure) { @T.0 }
+public fn f(@Nat -> @Int) requires(true) ensures(true) effects(pure) { let @Int = identity(@Nat.0); @Int.0 }
+""", "f"),
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_int_widening_differential.py:153]: @coderabbitai Keeping these — they pass. The callee is monomorphised to `identity$Nat` at the `@Nat` call site, so the bare-`T` lowering concern does not apply: `identity$Nat` lowers and runs fine, and both cases trap on `u64.MAX` (they returned `-1` before this PR, which is exactly the verifier↔codegen desync this corpus must catch — a `@Nat`-returning call result the verifier marks `tier3` but codegen left unguarded). Removing them would drop the only differential coverage of the call-result widening class, which was the headline finding of this review round.

**Comment by aallan** [comment]
@coderabbitai The two outside-diff TESTING.md findings are both addressed in 54b81fb — added `check_doc_builtin_shadowing.py` to the validation-scripts inventory (and bumped Twenty → Twenty-one) and fixed the conformance-suite prose count (102 → 103).

**Comment by aallan** [comment]
@coderabbitai review

File: scripts/check_doc_builtin_shadowing.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    ``_EXEMPT`` reference files and the generated / vendored / artefact trees in
+    ``_SKIP_DIRS`` are excluded."""
+    files = sorted(root.rglob("*.md"))
+    return [
+        f
+        for f in files
+        if f.relative_to(root).as_posix() not in _EXEMPT
+        and not any(
+            part in _SKIP_DIRS for part in f.relative_to(root).parts
+        )
```
[the remark this answers is not in the document]
Comment (aallan) [reply scripts/check_doc_builtin_shadowing.py]: @coderabbitai Fixed — `doc_files()` now uses `os.walk` and prunes `_SKIP_DIRS` from `dirnames` in place, so it never descends into `.venv` / `node_modules` / `docs` (the `rglob` version stat-ed everything first). Same 45-file coverage; gate still passes.

File: tests/test_doc_builtin_shadowing.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    """A `forall<...> fn <builtin>` generic header still redefines the built-in
+    and would E151, so the regex must look past the generic header — including a
+    `where`-bounded one with nested `<...>` (CR #821 review)."""
+    reject = reject_names()
+    assert find_shadowing_defs(
+        "public forall<T> fn abs(@T -> @T)\n", reject) == [(1, "abs")]
+    assert find_shadowing_defs(
+        "private forall<T where Eq<T>> fn clamp(@T -> @T)\n", reject) == [
+        (1, "clamp")
+    ]
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_doc_builtin_shadowing.py:68]: @coderabbitai Fixed — the regression test now also asserts a genuinely nested header with multiple `>` pairs: `forall<T where Ord<Map<K, V>>> fn max(...)`, so the greedy `forall<.*>` is forced past every inner `>`.

**Comment by aallan** [comment]
@coderabbitai The TESTING.md CI lint-job inventory finding is addressed in f553de1 — added `check_doc_builtin_shadowing.py` to the **lint** job script list (it runs at ci.yml:153), so it now matches the validation-scripts table.

**Comment by aallan** [comment]
@coderabbitai review
